buildSrc/src/main/java/org/elasticsearch/gradle/testclusters/RateLimitInternalExtension.java

```diff
@@ -0,0 +1,98 @@
+package org.elasticsearch.gradle.testclusters;
+
+import org.gradle.api.Project;
+import org.gradle.api.logging.Logger;
+import org.gradle.api.logging.Logging;
+
+import java.util.Optional;
+import java.util.UUID;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.Semaphore;
+import java.util.concurrent.TimeUnit;
+
+/**
+ * This extensions was meant to be used internally by testclusters
+ *
+ * It holds synchronization primitives needed to implement the rate limiting.
+ * This is tricky because we can't use Gradle workers as there's no way to make sure that tests and their clusters are
+ * allocated atomically, so we could be in a situation where all workers are tests waiting for clusters to start up.
+ *
+ * Also auto configures cleanup of executors to make sure we don't leak threads in the daemon.
+ */
+public class RateLimitInternalExtension {
+
+    private static final int EXECUTOR_SHUTDOWN_TIMEOUT = 1;
+    private static final TimeUnit EXECUTOR_SHUTDOWN_TIMEOUT_UNIT = TimeUnit.MINUTES;
+
+    private static final Logger logger =  Logging.getLogger(RateLimitInternalExtension.class);
+
+    private final Semaphore globalSemaphore;
+    private final ExecutorService executorService;
+    private final TestClusterCleanupOnShutdown cleanupThread;
+
+    public RateLimitInternalExtension() {
+        globalSemaphore = new Semaphore(maxPermits());
+        executorService = Executors.newFixedThreadPool(
```

Comment:
There's no need to construct a fixed thread pool here since it's the `Semaphore` that is the thing thats limiting concurrency, not the thread pool size. We'd be better off using an unbounded cached thread pool.